File: src/year2015/day03/parse.cc
#include "parse.h"

#include <stdexcept>
#include <string>
auto Year2015::Day03::parse(const std::string& input) -> Directions {
  Directions directions;
  directions.reserve(input.size());
  for (const auto& chr : input) {
    switch (chr) {
      case '^': {
        directions.emplace_back(Direction::North);
        break;
      }
      case 'v': {
        directions.emplace_back(Direction::South);
        break;
      }
      case '<': {
        directions.emplace_back(Direction::West);
        break;
      }
      case '>': {
        directions.emplace_back(Direction::East);
        break;
      }
      default: {
        throw std::runtime_error("Unexpected character '" +
                                 std::to_string(chr) +
                                 "' when parsing input.txt");
      }
    }
  }
  return directions;
}

File: src/year2015/day03/parse.rs
#[derive(Debug)]
pub enum Direction {
  North,
  South,
  East,
  West,
}

pub type Directions = Vec<Direction>;

pub fn parse(input: String) -> Result<Directions, String> {
  let mut directions = Directions::new();
  directions.reserve(input.len());
  for chr in input.bytes() {
    match chr {
      b'^' => {
        directions.push(Direction::North);
      }
      b'v' => {
        directions.push(Direction::South);
      }
      b'<' => {
        directions.push(Direction::East);
      }
      b'>' => {
        directions.push(Direction::West);
      }
      _ => {
        return Err(format!(
          "Unexpected character '{}' when parsing input.txt",
          chr
        ));
      }
    }
  }
  return Ok(directions);
}
